### PR TITLE
update(parser) - ELEXON

### PR DESCRIPTION
**Pull request by mathilde-daugy**
## Description
Here is a diagram of how the data is collected 
gb parser

We assume that the hydro storage from ESO is only pumping (from the grid to the reservoir) because it happens in the night and B1620 is discharge (from the reservoir to the grid) because it happens during the day at what see to be peak hours
image

This means that storage from B1620 is counted as negative and storage from ESO is counted as >0

### Double check

- [ ] I have tested my parser changes locally with `poetry run test_parser "zone_key"`
- [ ] I have run `pnpx prettier@2 --write .` and `poetry run format` in the top level directory to format my changes.

### CODE REVIEW

**Comment by VIKTORVAV99** [comment]
Have you seen this new Elexon API? https://developer.data.elexon.co.uk/api-details#api=prod-insol-insights-api&operation=get-datasets-fuelinst

I have not looked through them all yet but there could be some nice gems in there.

**Comment by mathilde-daugy** [comment]
> Have you seen this new Elexon API? [developer.data.elexon.co.uk/api-details#api=prod-insol-insights-api&operation=get-datasets-fuelinst](https://developer.data.elexon.co.uk/api-details#api=prod-insol-insights-api&operation=get-datasets-fuelinst)
> 
> I have not looked through them all yet but there could be some nice gems in there.

yes actually the revamp of the parser uses the new API. It's just that i still need to look at the pumping data before  the PR is ready for review

File: parsers/ELEXON.py
Comment (VIKTORVAV99) [inline parsers/ELEXON.py:131]:
```suggestion
```

File: parsers/ELEXON.py
Comment (VIKTORVAV99) [inline parsers/ELEXON.py:410]:
```suggestion
```

File: parsers/ELEXON.py
Comment (VIKTORVAV99) [inline parsers/ELEXON.py:287]:
```suggestion
) -> list[dict[str, Any]]:
```

File: parsers/ELEXON.py
Comment (VIKTORVAV99) [inline parsers/ELEXON.py:276]:
```suggestion
        storage_mix.add_value("hydro", float(event.get("PUMP_STORAGE_PUMPING")) if event.get("PUMP_STORAGE_PUMPING") else None)
```
Might make sense to extract it to a temp variable to avoid calling get twice but it's not needed.

File: parsers/ELEXON.py
Comment (VIKTORVAV99) [inline parsers/ELEXON.py:272]: While it never should be none we need to make handle it to satisify the typing (with `isinstance` for example).

File: parsers/ELEXON.py
Comment (VIKTORVAV99) [inline parsers/ELEXON.py:213]:
```suggestion
) -> list[dict[str, Any]]:
```

File: parsers/ELEXON.py
Comment (VIKTORVAV99) [inline parsers/ELEXON.py:173]: This also needs to be casted to float and asserted that it's not none. Might make sense to create a helper function for that since we use it more than once.

File: parsers/ELEXON.py
Comment (VIKTORVAV99) [inline parsers/ELEXON.py:163]: And this need to be type wrangled as well.

File: parsers/ELEXON.py
Comment (VIKTORVAV99) [inline parsers/ELEXON.py:137]:
```suggestion
    production_data: list[dict[str, Any]], logger: Logger, dataset: str
```

File: parsers/ELEXON.py
Comment (VIKTORVAV99) [inline parsers/ELEXON.py:121]:
```suggestion
) -> ProductionBreakdownList:
```

**Review by VIKTORVAV99 — COMMENTED** [review]
Lots of small notes that needs to be addressed.

I'm also seeing a lot of these in the terminal:
```
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
WARNING:test_parser:GB reported total of 0.00MW falls outside range of (100, 60000) for key `gas`
```
Is this expected? I don't see why they would report 0 gas (and nuclear).

**Comment by mathilde-daugy** [comment]
will remove this parser validation. I don't think it makes a lot of sese

**Comment by VIKTORVAV99** [comment]
I'll look into this further but there seems to be something seriously wrong with the earlier data:
```python
{'correctedModes': [],
  'datetime': datetime.datetime(2024, 2, 19, 0, 0, tzinfo=datetime.timezone.utc),
  'production': {'biomass': 0.0,
                 'coal': 0.0,
                 'gas': 0.0,
                 'hydro': 0.0,
                 'nuclear': 0.0,
                 'oil': 0.0,
                 'solar': 0.0,
                 'unknown': 0.0,
                 'wind': 10108.67},
  'source': 'elexon.co.uk, nationalgrideso.com',
  'sourceType': <EventSourceType.measured: 'measured'>,
  'storage': {'hydro': 11.0},
  'zoneKey': 'GB'}
```
(This is after I changed so 0 is not considered false so we get 0 solar at night).
I think this stems from the fact that we use two different endpoints? And therefore there is a misalignment?

File: parsers/ELEXON.py
Comment (VIKTORVAV99) [inline parsers/ELEXON.py:422]: This does not work and break it all basically, first of all the first data always have a length cause it always have wind production but the rest of the values are 0.

File: parsers/ELEXON.py
Comment (VIKTORVAV99) [inline parsers/ELEXON.py:334]: This should not return a list, but the merged events directly, it should also check that all datetimes are mathcing.

File: parsers/ELEXON.py
Comment (VIKTORVAV99) [inline parsers/ELEXON.py:429]: Here we are once again merging these datapoints that was merged in the `query_and_merge_production_fuelhh_and_eso` function, effectivly merging them twice.

File: parsers/ELEXON.py
Comment (VIKTORVAV99) [inline parsers/ELEXON.py:133]: What do we even use this data for? Why not use the fuel HH data only? that would drastically simplify the parser (and reduce error risks).

**Review by VIKTORVAV99 — CHANGES_REQUESTED** [review]
If I go to https://data.elexon.co.uk/bmrs/api/v1/datasets/FUELHH/stream?publishDateTimeFrom=2024-02-10&publishDateTimeTo=2024-02-11 everything seems to be fine data wise.

But if I run `poetry run test_parser GB --target_datetime="2024-02-11"` I get a bunch of zero filled data which indicated there is something flawed in either the datasets or our logic.

Turns out it's probably both. I have adressed them in individual comments.

File: parsers/ELEXON.py
Comment (VIKTORVAV99) [inline parsers/ELEXON.py:190]: This if clause is what is messing with the solar values since 0 is false but it's also filtering out all the 0 mess from the B1620 dataset.

**Comment by q--** [comment]
Adding a link to #6419 so its relationship to this pull request is clear in GitHub (though I'm unclear if this pull request addresses the issue or that it was merely looked into.)

File: parsers/ELEXON.py
Comment (mathilde-daugy) [reply parsers/ELEXON.py:133]: yes but HH does not include hydro storage and that is the reason why we had the original issue. the B1620 dataset splits hydro in hydro + hydro discharge.

Comment (mathilde-daugy) [reply parsers/ELEXON.py:133]: the B1620 data uses the same reporting convention as ENTSOE

Comment (mathilde-daugy) [reply parsers/ELEXON.py:133]: This endpoint provides actual generation data per settlement period aggregated by Power System Resource type (Fuel Type categories as defined by Commission Regulation (EU) No 543/2013)

**Comment by mathilde-daugy** [comment]
converting back to draft, i will address this next week.

File: parsers/ELEXON.py
Comment (mathilde-daugy) [reply parsers/ELEXON.py:133]: also the HH data is incomplete for wind (it doesn't include distributed) and we need to merge it with the ESO data so it is not really that straight forward. 
The best is to use the B1620 data when it is available and use the other data as backup

Comment (mathilde-daugy) [reply parsers/ELEXON.py:133]: see #4899

File: parsers/ELEXON.py
Comment (mathilde-daugy) [reply parsers/ELEXON.py:429]: no that is not what is happening here. We are only collecting hydro storage data from the eso source and merging it with the B1620 data that only has discharge so it is not double counting

File: parsers/ELEXON.py
Comment (mathilde-daugy) [reply parsers/ELEXON.py:334]: it should actually return a list because it will not be merged with anything else

File: parsers/ELEXON.py
Comment (mathilde-daugy) [reply parsers/ELEXON.py:190]: I've added a validation function further down that checks if the production breakdown is complete

**Comment by mathilde-daugy** [comment]
> Adding a link to #6419 so its relationship to this pull request is clear in GitHub (though I'm unclear on whether this pull request addresses the issue, or that it was merely looked into).

thanks for the link. This is indeed an answer to the issue you raised about discharging at night :) 

From my investigation, it seems that ESO data only covers pumping and not discharge and that is why the data is off, it is missing the other half of the puzzle!

**Comment by mathilde-daugy** [comment]
image

File: parsers/ELEXON.py
Comment (VIKTORVAV99) [inline parsers/ELEXON.py:248]:
```suggestion
    production_data: list[dict[str, Any]], logger: Logger
```

File: parsers/ELEXON.py
Comment (VIKTORVAV99) [inline parsers/ELEXON.py:290]:
```suggestion
    eso_data: list[dict[str, Any]], logger: Logger
```

File: parsers/ELEXON.py
Comment (VIKTORVAV99) [reply parsers/ELEXON.py:197]: Does that mean this can be removed?

It's still causing the following issue with no solar at night:
```python
 {'correctedModes': [],
  'datetime': datetime.datetime(2024, 3, 27, 5, 30, tzinfo=datetime.timezone.utc),
  'production': {'biomass': 1816.0,
                 'coal': 270.0,
                 'gas': 2938.0,
                 'hydro': 345.0,
                 'nuclear': 4564.0,
                 'unknown': 418.0,
                 'wind': 14942.188},
  'source': 'elexon.co.uk, nationalgrideso.com',
  'sourceType': <EventSourceType.measured: 'measured'>,
  'storage': {'hydro': -135.0},
  'zoneKey': 'GB'},
 {'correctedModes': [],
  'datetime': datetime.datetime(2024, 3, 27, 6, 0, tzinfo=datetime.timezone.utc),
  'production': {'biomass': 2086.0,
                 'coal': 329.0,
                 'gas': 3589.0,
                 'hydro': 380.0,
                 'nuclear': 4582.0,
                 'solar': 78.0,
                 'unknown': 425.0,
                 'wind': 15143.47},
  'source': 'elexon.co.uk, nationalgrideso.com',
  'sourceType': <EventSourceType.measured: 'measured'>,
  'storage': {'hydro': 65.0},
  'zoneKey': 'GB'},
  ```

File: parsers/ELEXON.py
Comment (VIKTORVAV99) [inline parsers/ELEXON.py:361]: There is no guarantee that get is available since the response is a JSON response, we need to ensure it is a dict before calling get on it.

The type data also says it will return a list, which causes type issues.

**Review by VIKTORVAV99 — COMMENTED** [review]
Seems to be working but there is still the issue of no data at night (it's missing when it should be 0).

File: parsers/ELEXON.py
Comment (VIKTORVAV99) [reply parsers/ELEXON.py:364]: This is just a type issue so ignoring it for now since we need to get this merged ASAP.

**Review by VIKTORVAV99 — APPROVED** [review]
I did some small changes and now the output looks a bit better with less errors by correcting negative values with 0 (so they are still valid) and by only merging the matching timestamps (so we ensure we get complete data before the validation method kicks in).

There might still be an issue with no solar at all during the night (not null or 0, simply missing) but it's not blocking from my side. So I'll leave that up to you.

I'm going to bussy for the rest of this and next week so I won't be able to fix that particular issue if you consider it blocking.

**Comment by VIKTORVAV99** [comment]
@tonypls @madsnedergaard could you just give this a quick double check?
I know there is a lot of changes and they should all be fine but it would be nice to get another set of eyes on it. We need to merge this before the 1st so it's kind of urgent as they will be shutting down the old API.

So I'll be merging this tomorrow no matter what so we can monitor it during Friday if needed.

**Review by madsnedergaard — APPROVED** [review]
LGTM, thanks both of you for the work here ❤️

**Pull request merged**
